### PR TITLE
210 209 add qr code join

### PR DESCRIPTION
**Pull request by Flamantrose321**

### CODE REVIEW

File: app/src/androidTest/java/com/github/fribourgsdp/radio/PlayerGameHandlerTest.kt [hunk 1]
```diff
@@ -256,7 +256,7 @@ class PlayerGameHandlerTest {
         handler.onPick(song)
 
         // Wait for the task of the database to execute
-        Thread.sleep(1)
+        Thread.sleep(100)
```
Comment (BartLucien) [inline app/src/androidTest/java/com/github/fribourgsdp/radio/PlayerGameHandlerTest.kt:259]: Why did you need to extend to sleeping time?

File: app/src/androidTest/java/com/github/fribourgsdp/radio/mockimplementations/MockGameActivity.kt [hunk 2]
```diff
@@ -1,11 +1,15 @@
 package com.github.fribourgsdp.radio.mockimplementations
 
 import com.github.fribourgsdp.radio.*
+import io.agora.rtc.IRtcEngineEventHandler
+import io.agora.rtc.RtcEngine
+import org.mockito.Mockito
```
Comment (VictorSchneuwly) [inline app/src/androidTest/java/com/github/fribourgsdp/radio/mockimplementations/MockGameActivity.kt:6]: Do you still need those imports ?

File: app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        // initializing onclick listener for button.
+        generateQRCode()
+
+    }
+
+    private fun generateQRCode(){
+
+        if (TextUtils.isEmpty(lobbyId.toString())) {
```
Comment (VictorSchneuwly) [inline app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt:62]: Since `lobbyId` is a `Long`, how could this check be empty?

File: app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+
+class CreateQRCodeFragment(val ctx: Context, val lobbyId: Long ): DialogFragment() {
+    // variables for imageview, edittext,
+    // button, bitmap and qrencoder.
+    private lateinit var qrCodeIV: ImageView
+    private lateinit var rootView : View
+
+    private lateinit var cancelButton: Button
+    lateinit var bitmap: Bitmap
```
Comment (VictorSchneuwly) [inline app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt:36]: I think `bitmap`doesn't need to be extracted.

File: app/src/androidTest/java/com/github/fribourgsdp/radio/PlayerGameHandlerTest.kt (hunk 1 above)
Comment (Flamantrose321) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/PlayerGameHandlerTest.kt:259]: The test failed several times when I pushed, so I added time to make it pass every time. I notified Victor so that he could look into optimising this.

File: app/src/androidTest/java/com/github/fribourgsdp/radio/mockimplementations/MockGameActivity.kt (hunk 2 above)
Comment (Flamantrose321) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/mockimplementations/MockGameActivity.kt:6]: you are right, they are useless. thank you

File: app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt (hunk 3 above)
Comment (Flamantrose321) [reply app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt:62]: This was an additional security, but it is indeed not very necessary.

File: app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt (hunk 4 above)
Comment (Flamantrose321) [reply app/src/main/java/com/github/fribourgsdp/radio/CreateQRCodeFragment.kt]: Yes, you are right

File: app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingJoinGameActivityTest.kt [hunk 5]
```diff
@@ -47,42 +56,113 @@ class WorkingJoinGameActivityTest {
...
+            joinQRCodeButton.perform(ViewActions.click())
+
+            Espresso.onView(ViewMatchers.withId(R.id.cancel_button))
+                .inRoot(RootMatchers.isDialog())
+                .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
+
+            Espresso.onView(ViewMatchers.withId(R.id.cancel_button))
+                .inRoot(RootMatchers.isDialog())
+                .perform(ViewActions.click())
+        }
```
Comment (BartLucien) [inline app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingJoinGameActivityTest.kt:119]: Don't you want to test that what happens when you click on cancel works correctly?

File: app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingLobbyActivityTest.kt [hunk 6]
```diff
@@ -166,46 +151,72 @@ class WorkingLobbyActivityTest {
...
+            val displayQRCodeButton = Espresso.onView(withId(R.id.displayQRCodeButton))
+            displayQRCodeButton.perform(ViewActions.click())
+
+            Espresso.onView(withId(R.id.cancel_button))
+                .inRoot(isDialog())
+                .check(matches(isDisplayed()))
+
+            Espresso.onView(withId(R.id.cancel_button))
+                .inRoot(isDialog())
+                .perform(ViewActions.click())
```
Comment (BartLucien) [inline app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingLobbyActivityTest.kt:217]: Same remark as above, don't you want to test what it does after cancelling?

File: app/src/main/java/com/github/fribourgsdp/radio/JoinWithQRCodeFragment.kt [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            Manifest.permission.VIBRATE
+        ), PERMISSION_REQUEST_CODE)
+    }
+
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        // this method is called when user
+        // allows the permission to use camera.
+        if (grantResults.size > 0) {
+            val cameraaccepted = grantResults[0] == PackageManager.PERMISSION_GRANTED
```
Comment (BartLucien) [inline app/src/main/java/com/github/fribourgsdp/radio/JoinWithQRCodeFragment.kt:154]: I think you should rename the variable name with as cameraAccepted. Same for below variables

File: app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingJoinGameActivityTest.kt (hunk 5 above)
Comment (Flamantrose321) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingJoinGameActivityTest.kt:119]: yes indeed an oversight on my part, well done

File: app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingLobbyActivityTest.kt (hunk 6 above)
Comment (Flamantrose321) [reply app/src/androidTest/java/com/github/fribourgsdp/radio/WorkingLobbyActivityTest.kt:217]: yes indeed an oversight on my part, well done

File: app/src/main/java/com/github/fribourgsdp/radio/JoinWithQRCodeFragment.kt (hunk 7 above)
Comment (Flamantrose321) [reply app/src/main/java/com/github/fribourgsdp/radio/JoinWithQRCodeFragment.kt]: yes indeed

**Review by BartLucien — APPROVED** [review]
Seems good to me

**Review by VictorSchneuwly — APPROVED** [review]
Good to me now 👍🏻
